edward/inferences/hmcda.py

```diff
@@ -0,0 +1,330 @@
+
+    # Sample momentum.
+    old_r = OrderedDict()
+
+    for z, qz in six.iteritems(self.latent_vars):
+      event_shape = qz.event_shape
+      normal = Normal(loc=tf.zeros(event_shape), scale=tf.ones(event_shape))
+      old_r[z] = normal.sample()
+
+    # Initialize espilon at 1.0
+    epsilon = tf.Variable(1.0, trainable=False)
+
+    # Calculate log joint probability
+    old_z = {z: tf.gather(qz.params, 0)
+             for z, qz in six.iteritems(self.latent_vars)}
+    old_z = OrderedDict(old_z)
+
+    log_p_joint = -kinetic_energy(old_r)
+    log_p_joint += self._log_joint(old_z)
+
+    new_sample, new_r_sample = leapfrog(old_z, old_r,
+                                        epsilon, self._log_joint, 1)
+
+    log_p_joint_prime = -kinetic_energy(new_r_sample)
+    log_p_joint_prime += self._log_joint(new_sample)
+
+    log_p_joint_diff = log_p_joint_prime - log_p_joint
+
+    # See whether epsilon is too small or to big
+    condition = log_p_joint_diff >= tf.log(0.5)
+    a = 2.0 * tf.where(condition, 1.0, 0.0) - 1.0
+
+    # Save keys of (z, r) so that we can rebuild the Dict inside the while_loop
+    keys_r = list(six.iterkeys(old_r))
+    keys_z = list(six.iterkeys(old_z))
+
+    k = tf.constant(0)
+
+    def while_condition(k, _, log_p_joint_prime_loop, log_p_joint, *args):
+      accep_big_enough = tf.pow(tf.exp(
+          log_p_joint_prime_loop - log_p_joint), a) > tf.pow(2.0, -a)
+      to_many_iterations = k < 12
+      return tf.logical_and(to_many_iterations, accep_big_enough)
+
+    def body(k, epsilon_loop, _, log_p_joint, values_r, values_z):
+        new_epsilon_loop = tf.pow(2.0, a) * epsilon_loop
+
+        # Rebuild the Dicts inside the while_loop since we can only return lists
+        old_z_loop = OrderedDict()
+        old_r_loop = OrderedDict()
+        for i, key in enumerate(values_z):
+          old_z_loop[keys_z[i]] = values_z[i]
+          old_r_loop[keys_r[i]] = values_r[i]
+
+        new_z_loop, new_r_loop = leapfrog(old_z_loop, old_r_loop,
+                                          new_epsilon_loop, self._log_joint, 1)
+        new_log_p_joint_prime = -kinetic_energy(new_r_loop)
+        new_log_p_joint_prime += self._log_joint(new_z_loop)
+
+        return [k + 1, new_epsilon_loop, new_log_p_joint_prime,
+                log_p_joint, values_r, values_z]
+
+    _, new_epsilon, _, _, _, _ = tf.while_loop(
+        while_condition, body,
+        loop_vars=[k, epsilon, log_p_joint_prime, log_p_joint,
+                   list(six.itervalues(old_r.copy())),
+                   list(six.itervalues(old_z.copy()))])
+
+    return new_epsilon
+
+  def _log_joint(self, z_sample):
+    """Utility function to calculate model's log joint density,
+    log p(x, z), for inputs z (and fixed data x).
+
+    Parameters
+    ----------
+    z_sample : dict
+      Latent variable keys to samples.
+    """
+    self.scope_iter += 1
+    scope = 'inference_' + str(id(self)) + '/' + str(self.scope_iter)
```

Comment:
Note we no longer use `scope_iter` and `str(id(self))` implementation for scopes. See the latest versions of `hmc.py` and `sghmc.py` where we use `scope = tf.get_default_graph().unique_name("inference")`.